<a href="https://colab.research.google.com/github/zamoralinokevin-eng/Maestria-en-IA/blob/main/Actividad_Calificada_07_Datos_Temporales_271753.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | 2026-09-25 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** Kevin Josue Zamora Lino | |
| **Matrícula** 271753 | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [2]:
#se muestra el tipo de dato de pickup and dropoff
print("El tipo de dato de pickup es:",taxis["pickup"].dtype)
print("El tipo de dato de dropoff es:", taxis["dropoff"].dtype)

#se verifica los NaT generados antes de la conversion
nulos_pickup_antes = taxis["pickup"].isna().sum()
nulos_dropoff_antes = taxis["dropoff"].isna().sum()
print("\nNúmero de NaT en pickup antes de la conversión:", nulos_pickup_antes)
print("Número de NaT en dropoff antes de la conversión:", nulos_dropoff_antes)

#se comvierten ambas columnas a tipo fecha indicando el formato
taxis["pickup"] = pd.to_datetime(taxis["pickup"], format="%Y-%m-%d %H:%M:%S")
taxis["dropoff"] = pd.to_datetime(taxis["dropoff"], format="%Y-%m-%d %H:%M:%S")

#se verifica NaT generados despues de la conversion
nulos_pickup_despues = taxis["pickup"].isna().sum()
nulos_dropoff_despues = taxis["dropoff"].isna().sum()
print("\nNúmero de NaT en pickup después de la conversión:", nulos_pickup_despues)
print("Número de NaT en dropoff después de la conversión:", nulos_dropoff_despues)

El tipo de dato de pickup es: object
El tipo de dato de dropoff es: object

Número de NaT en pickup antes de la conversión: 0
Número de NaT en dropoff antes de la conversión: 0

Número de NaT en pickup después de la conversión: 0
Número de NaT en dropoff después de la conversión: 0


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [3]:
#se muestran las fechas minimas y maximas de pickup
print("Fecha mínima de pickup:", taxis["pickup"].min())
print("Fecha máxima de pickup:", taxis["pickup"].max())

#se muestran las fechas minimas y maximas de dropoff
print("\nFecha mínima de dropoff:", taxis["dropoff"].min())
print("Fecha máxima de dropoff:", taxis["dropoff"].max())

#se filtran todos los viajes que no sean de marzo 2019
viajes_fuera_de_marzo = taxis[~((taxis["pickup"].dt.year == 2019)&(taxis["pickup"].dt.month == 3))]
print("\n\nLos viajes fuera de Marzo 2019 son", len(viajes_fuera_de_marzo))
print("\n(Se muestra viajes fuera de marzo)\n", viajes_fuera_de_marzo[['pickup', 'dropoff']])


viajes_pickup_dropoff_fuera = taxis[
    (taxis['pickup'].dt.month == 3) &~(taxis['dropoff'].dt.month == 3)]

print("\n\nLos viajes con dropoff fuera de marzo son:", len(viajes_pickup_dropoff_fuera))
print("\n(Se muestra viajes con dropoff fuera de marzo)\n",
      viajes_pickup_dropoff_fuera[['pickup', 'dropoff']])



Fecha mínima de pickup: 2019-02-28 23:29:03
Fecha máxima de pickup: 2019-03-31 23:43:45

Fecha mínima de dropoff: 2019-02-28 23:32:35
Fecha máxima de dropoff: 2019-04-01 00:13:58


Los viajes fuera de Marzo 2019 son 1

(Se muestra viajes fuera de marzo)
                   pickup             dropoff
6203 2019-02-28 23:29:03 2019-02-28 23:32:35


Los viajes con dropoff fuera de marzo son: 1

(Se muestra viajes con dropoff fuera de marzo)
                  pickup             dropoff
591 2019-03-31 23:43:45 2019-04-01 00:13:58


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [4]:
#se filtran los viajes con pickup 10 de marzo
viajes_10_marzo = taxis[taxis['pickup'].dt.strftime("%Y-%m-%d")=="2019-03-10"]

#se cuenta cuantos hay en cada hora
viajes_por_hora = viajes_10_marzo.groupby(viajes_10_marzo['pickup'].dt.hour).size()
print("Viajes por hora:\n", viajes_por_hora)
print(f"¿Hay viajes a las 2:00 AM?: {'Sí' if 2 in viajes_por_hora.index else 'No'}")

#se localiza pickup en la zona America/New_York usando noneistent="NaT"
pickup_localizado = taxis["pickup"].dt.tz_localize("America/New_York", nonexistent="NaT")

#se cuenta los valores que quedaron como NaT
nuevos_nat = pickup_localizado.isna().sum() - taxis["pickup"].isna().sum()
print(f"\nLos registros transformados a NaT debido al cambio de horario son {nuevos_nat}")


Viajes por hora:
 pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
dtype: int64
¿Hay viajes a las 2:00 AM?: No

Los registros transformados a NaT debido al cambio de horario son 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

_Escribe aquí tu respuesta._ Por los analisis que se han hecho hasta el momento, el dataset esta limpio, sus columnas tienen un formato establecido de datetime. El analisis mostró que no habian NaT (0 registros) antes y despues de la conversion de pickup y dropoff. En la evaluacion del punto 1.3 no se encontraron registros (0 datos) transformados debido al cambio de horario, es decir, se saltó de la hora 1 a la 3. Tampoco existen viajes registrados a las 2am.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [5]:
#se crea la columna taxis["duracion"]
taxis["duracion"] = taxis["dropoff"] - taxis["pickup"]

#se crea la columna taxis["duracion_min"]
taxis["duracion_min"] = taxis["duracion"].dt.total_seconds() / 60

#se muestra el resumen estadistico de la columna taxis["duracion_min"]
print("Resumen estadístico de 'duracion_min':\n", taxis["duracion_min"].describe())


Resumen estadístico de 'duracion_min':
 count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [6]:
#se crea una mascara para duracion negativa y se imprime resultado
duracion_negativa = taxis[taxis["duracion_min"]<0]
print("Viajes con duracion negativa:", len(duracion_negativa))

#se crea una mascara para duracion cero y se imprime resultado
duracion_cero = taxis[taxis["duracion_min"]==0]
print("Viajes con duracion cero:", len(duracion_cero))

#se crea una mascara con duracion menora 1 minuto y se imprime resultado
duracion_menor_a_1 = taxis[taxis["duracion_min"]<1]
print("Viajes con duracion menor a 1 minuto:", len(duracion_menor_a_1))

#se crea una mascara con duracion mayor a 3 horas y se imprime resultado
duracion_mayor_a_3 = taxis[taxis["duracion_min"]>180]
print("Viajes con duracion mayor a 3 horas:", len(duracion_mayor_a_3))

#se muestran los viajes de duracion cero con las columnas arriba citadas
columnas_deseadas = ["pickup", "dropoff", "distance", "fare", "total"]
print("\nViajes con duracion cero con las columnas deseadas:\n", duracion_cero[columnas_deseadas])

Viajes con duracion negativa: 0
Viajes con duracion cero: 6
Viajes con duracion menor a 1 minuto: 72
Viajes con duracion mayor a 3 horas: 0

Viajes con duracion cero con las columnas deseadas:
                   pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [7]:
#se crea la columna velocidad_mph para viajes > 0
taxis["velocidad_mph"] = taxis["distance"] / (taxis["duracion_min"] / 60).where(taxis["duracion_min"]>0)

#se cuentan los viajes con velocidad >60 mph
viajes_mayores_60 = taxis[taxis["velocidad_mph"]>60]
print("Viajes con velocidad mayor a 60 mph:", len(viajes_mayores_60))

#se muestran los 5 viajes con mayor velocidad, ordenados de mayor a menor
columnas_a_mostrar = ["pickup", "dropoff", "duracion_min", "distance", "velocidad_mph", "fare"]
viajes_mayores_60_sorted = viajes_mayores_60.sort_values(by="velocidad_mph", ascending=False)
print("\n\nTop 5 viajes con mayor velocidad:\n", viajes_mayores_60_sorted[columnas_a_mostrar].head(5))

Viajes con velocidad mayor a 60 mph: 11


Top 5 viajes con mayor velocidad:
                   pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

_Escribe aquí tu respuesta._ Los viajes que excluiria del analisis son los de velocidad mayor a 60 mph porque las velocidades registradas no son reales para ningun tipo de auto. Los viajes con tiempo < 1 minuto no son errores porque registran tarifas cobradas, significando que es probable que se haya cancelado el viaje por diferentes motivos, Este mismo tipo de viaje presenta distancias de cero por la cancelacion del viaje.
Los umbrales que usaria son la velocidad ya que como mencionaba, existen valores registrados que no son alcanzables. Tambien usaria el pickup y dropoff para excuir viajes que no sean de Marzo 2019.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [8]:
#se crea la columna hora
taxis["hora"] = taxis["pickup"].dt.hour

#se crea la columna dia_num
taxis["dia_num"] = taxis["pickup"].dt.dayofweek

#se crea la columna dia_nombre
taxis["dia_nombre"] = taxis["pickup"].dt.day_name()

#se crea fin de semana si es sabado o domingo
taxis["fin_semana"] = taxis["dia_nombre"].isin(["Saturday", "Sunday"])

#se crea la columna fecha
taxis["fecha"] = taxis["pickup"].dt.date

#se crea la columna cruza_medianoche
taxis["cruza_medianoche"] = taxis["pickup"].dt.date != taxis["dropoff"].dt.date
print("Viajes que cruzan la medianoche:", taxis["cruza_medianoche"].sum())

Viajes que cruzan la medianoche: 50


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [9]:
#se calcula para cada hora el numero de viajes y mediana
patron_por_hora = taxis.groupby("hora").agg(
    numero_viajes = ("pickup", "count"),
    mediana_duracion = ("duracion_min", "median")
)

#se indica la hora con mas viajes y la hora con menos viajes
hora_mas_viajes = patron_por_hora["numero_viajes"].idxmax()
hora_menos_viajes = patron_por_hora["numero_viajes"].idxmin()

#se imprimen valores, incluyendo el uso de la mediana
print("\nPatrón de numero de viajess por hora del día:")
print(patron_por_hora)
print(f"\nHora con mas viajes: {hora_mas_viajes}h con {patron_por_hora.loc[hora_mas_viajes, 'numero_viajes']} viajes.")
print(f"Hora con menos viajes: {hora_menos_viajes}h con {patron_por_hora.loc[hora_menos_viajes, 'numero_viajes']} viajes.")





Patrón de numero de viajess por hora del día:
      numero_viajes  mediana_duracion
hora                                 
0               205         10.733333
1               110          9.633333
2               101          9.183333
3                67          9.583333
4                57         10.500000
5                51          8.700000
6               142          7.833333
7               221          9.583333
8               313         11.316667
9               319         11.616667
10              325         11.883333
11              295         12.033333
12              334         12.775000
13              316         10.558333
14              360         11.841667
15              330         12.058333
16              336         11.583333
17              388         11.975000
18              417         11.100000
19              406         10.650000
20              368          9.258333
21              355         11.083333
22              321          9.816667
23 

### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [12]:
#se agrupan los datos  por dia_num y di_nombre
patron_por_dia = taxis.groupby(["dia_num", "dia_nombre"]).agg(
    numero_viajes = ("pickup", "count"),
    mediana_duracion = ("duracion_min", "median"),
    mediana_tarifa = ("fare", "median")
).reset_index()

#se calcula el promedio de viajes por fecha
fechas_por_dia = taxis.groupby(["dia_num", "dia_nombre"])["fecha"].nunique().reset_index(name='fechas_unicas')
patron_por_dia = patron_por_dia.merge(fechas_por_dia, on=["dia_num", "dia_nombre"])

#se calcula el promedio de viajes por fecha
patron_por_dia["promedio_viajes_por_fecha"] = patron_por_dia["numero_viajes"] / patron_por_dia["fechas_unicas"]

#se imprimer los resultados obtenidos
print("Análisis estandarizado por día de la semana")
print(patron_por_dia[['dia_nombre', 'numero_viajes', 'fechas_unicas', 'promedio_viajes_por_fecha', 'mediana_duracion', 'mediana_tarifa']])

Análisis estandarizado por día de la semana
  dia_nombre  numero_viajes  fechas_unicas  promedio_viajes_por_fecha  \
0     Monday            708              4                     177.00   
1    Tuesday            825              4                     206.25   
2  Wednesday            966              4                     241.50   
3   Thursday            905              5                     181.00   
4     Friday           1115              5                     223.00   
5   Saturday           1046              5                     209.20   
6     Sunday            868              5                     173.60   

   mediana_duracion  mediana_tarifa  
0         10.541667             9.5  
1         11.216667             9.5  
2         11.291667             9.5  
3         11.766667             9.5  
4         11.150000             9.5  
5         10.466667             9.0  
6          9.983333             9.0  


### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._Los patrones que se observan por hora son que a las 18-19 horas existe registrado el mayor numero de viajes, lo que es debido a que es la hora en que los ciudadanos salen del trabajo hacia sus casas con una mediana entre 10.65 y 11.1. Tambien los Viernes son los dias con la mayor cantidad de viajes registrados, siguiedo de este, el Sabado. Ambos dias es cuando las personas salen por diferentes motivos ajenos al trabajo. Las horas con mas viajes tienen una mediana mas grande debido a que a mayorcantidad de trafico, es mayor el tiempo de duracion del viaje debido al congestionamiento vehicular.
Fue necesario calcular el promedio de viajes por fecha para obtener  una metrica de densidad diaria comparable ya que si solo se hubiese considerado el total de viajes, este ocasionaria un sesgo de calendario ya que habrian mas dias de unos de que otros generando dias con mayor peso.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [20]:
#se aplican los filtros de la seccion 2.4, con condiciones por periodo, duracion y velocidad
condicion_periodo = (taxis["pickup"].dt.year == 2019) & (taxis["pickup"].dt.month == 3)
condicion_duracion = (taxis["duracion_min"] > 1)
condicion_velocidad = (taxis["velocidad_mph"] < 60)&(taxis["velocidad_mph"].notna())
taxis_limpio = taxis[condicion_periodo & condicion_duracion & condicion_velocidad]

#se cuantifican datos con los filtros creados
viajes_excluidos_periodo = len(taxis) - condicion_periodo.sum()
viajes_excluidos_duracion = len(taxis) - condicion_duracion.sum()
viajes_excluidos_velocidad = len(taxis[taxis["velocidad_mph"]>60])

#se crea el dataframe taxis_limpio
taxis_limpio = taxis[condicion_periodo & condicion_duracion & (taxis['velocidad_mph'] <= 60)].copy()
total_excluidos = len(taxis) - len(taxis_limpio)

print(f"Viajes excluidos por periodo (fuera de marzo 2019): {viajes_excluidos_periodo}")
print(f"Viajes excluidos por duración < 1 minuto: {viajes_excluidos_duracion}")
print(f"Viajes excluidos por velocidad > 60 mph: {viajes_excluidos_velocidad}")
print(f"Ttoal de registros excluidos: {total_excluidos}")
print(f"Dimensiones finales de 'taxis_limpio': {taxis_limpio.shape}")

Viajes excluidos por periodo (fuera de marzo 2019): 1
Viajes excluidos por duración < 1 minuto: 74
Viajes excluidos por velocidad > 60 mph: 11
Ttoal de registros excluidos: 76
Dimensiones finales de 'taxis_limpio': (6357, 23)


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [19]:
#se filtra el dataset con las horas entre 7 y 9
horas_7_9 = taxis_limpio['hora'].between(7, 9)

#se crea otro dataset con el filtro booleano
viajes_hora_7_9 = taxis_limpio[horas_7_9]

#se agrupa el nuevo dataset creado, se reporta mediana de duracion_min y velocidad_mph
comparativa = viajes_hora_7_9.groupby('fin_semana').agg(
    numero_viajes=('pickup', 'count'),
    mediana_duracion_min=('duracion_min', 'median'),
    mediana_velocidad_mph=('velocidad_mph', 'median')
).rename(index={False: 'Días Hábiles', True: 'Fin de Semana'})

print("Tabla comparativa de viajes con horario de 7 a 9:59\n")
print(comparativa)

Tabla comparativa de viajes con horario de 7 a 9:59

               numero_viajes  mediana_duracion_min  mediana_velocidad_mph
fin_semana                                                               
Días Hábiles             683             12.033333               9.097978
Fin de Semana            161              7.766667              13.473054


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._Primero se procedieron a cambiar el formato de las columnas pickup y dropoff a datetime junto con el formato de string con el que estaban guardadas. Se verificó los ausentes antes y despues de la conversion.
La zona horaria pertenece a America/New_York, dectectando que durante el cambio de horario en primavera, no habian datos a las 2am lo cual es un comportamiento normal con est tipo de zonas horarias.
Los criterios de exclusion de los datos fueron los registros de meses diferentes a Marzo, viajes con duracion menor a 1 minuto y velocidad mayor a 60 mph (los valores mayores a este ,eran irreales)
Como resultado del filtrado, se conservaron 6357 registros limpios para poder generar un modelo predictivo.
Como hallazgo, la mayoria de viajes en hora pico matutina (7-9:59) es entre semana (683 viajes) con medianas de duracion mas alta (12.03) y velocidades promedio mas bajas debido al trafic generado, en comparacion con los viajes el fin de semana (161 viajes con mediana de 7.76 de duracion). Lo cual resalta los viajes por motivos laborales en la semana en ese horario.


---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

_Escribe aquí tu respuesta._
Durante la exploracion inicial del dataset, si las fechas tienen un formato establecido pero tipo de dato diferente, es una buena practica establecer este formato en la conversion a datetime ya que facilita la conversion de tipo de dato y el procesamiento, aparte se evita que pandas infiera el formato con la primera fila que detecte y sea diferente a las otras con peligro de generar los errores silenciosos NaT cuando la fecha registrada no sea valida para el formato inferido.
El registro de estas fechas en diferentes zonas horarias generaria un desfase equivalente a la diferencia horaria mas el cambio de horario de verano dando como resultado complicaciones durante el calculo de tiempo transcurrido. Mediante el metodo describe() se identificaria esta diferencia par poder tomar decisiones previas de preprocesamiento de datos.
Toda anomalia temporal que no se limpie, va a generar ruido en el modelo predictivo pudiendo ser de alto impacto como las velocidades >300 mph registradas o duracion negativa o de cero por la difencia de zonas horarias. Por eso si no se limpian los datos antes de entrenar el modelo, este va a generar predicciones muy imprecisas y poco confiables.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.